### PR TITLE
DeLambda transformation

### PR DESCRIPTION
**Pull request by Kukovec**
- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] [Entries added to `./unreleased/`][changelog format] for any new functionality

Since #2532 is likely to require significant planning and refactoring, here's a temporary workaround for unsupported lambda-generated let applications.

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

File: tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/DeLambdifier.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import at.forsyte.apalache.tla.lir.transformations._
+
+/**
+ * DeLambdifier turns applications of operator-typed LET-expressions, into LET-expressions with applications pushed into
+ * the body.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+class DeLambdifier(tracker: TransformationTracker) extends TlaExTransformation {
```
Comment (shonfeder) [inline tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/DeLambdifier.scala:14]: I think `LetInApplier` is probably a more accurate name? There seems to already be precedent for `de-lambda` as a transformation that converts (proper) lambdas into let/in expressions, which is not what we're ding here :) https://en.wikipedia.org/wiki/Let_expression#Conversion_from_lambda_to_let_expressions

File: tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/DeLambdifier.scala [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package at.forsyte.apalache.tla.pp
+
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.lir.oper.TlaOper
+import at.forsyte.apalache.tla.lir.transformations._
+
+/**
+ * DeLambdifier turns applications of operator-typed LET-expressions, into LET-expressions with applications pushed into
+ * the body.
+ *
```
Comment (shonfeder) [inline tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/DeLambdifier.scala:10]: Maybe note that this is a workaround motivated by lack of support for lambdas on the IR, with a link to the issue?

**Review by shonfeder — APPROVED** [review]
Great, thanks! 
Two suggestions, tho they are not essential.

And, just to confirm, this rewriting pass will run on code that is supplied by to the checker via loading from JSON in https://github.com/informalsystems/apalache/blob/5463ad2c6913ddbdbbbb802a344da1d62a0bd708/passes/src/main/scala/at/forsyte/apalache/tla/passes/imp/SanyParserPassImpl.scala#L126

**Comment by shonfeder** [comment]
Or do i rather need to ensure i call this tranformer on every TlaEx constructed in the conversion?

**Comment by Kukovec** [comment]
> Or do i rather need to ensure i call this tranformer on every TlaEx constructed in the conversion?

We can adapt it a bit, and have it just be one of the transformations in `PreprocessingPass`, if that's easier, from a usability POV

**Comment by shonfeder** [comment]
That would definitely be much easier!

**Review by shonfeder — APPROVED** [review]
Awesome. This will be a big help. Thanks very much!!!

**Comment by shonfeder** [comment]
Unfortunately this workaround is not addressing the problem in quint conversion as hoped: 

```
    <unknown>: internal error in type checking: Bug in ToEtcExpr. Expected an operator name, found: LET __QUINT_LAMBDA0(x) ≜ x + 1 IN __QUINT_LAMBDA0 E@14:00:51.892
    at.forsyte.apalache.infra.AdaptedException: <unknown>: internal error in type checking: Bug in ToEtcExpr. Expected an operator name, found: LET __QUINT_LAMBDA0(x) ≜ x + 1 IN __QUINT_LAMBDA0
        at at.forsyte.apalache.infra.passes.PassChainExecutor$.run(PassChainExecutor.scala:39)
        at at.forsyte.apalache.tla.tooling.opt.CheckCmd.run(CheckCmd.scala:132)
        at at.forsyte.apalache.tla.Tool$.runCommand(Tool.scala:138)
        at at.forsyte.apalache.tla.Tool$.run(Tool.scala:118)
        at at.forsyte.apalache.tla.Tool$.main(Tool.scala:40)
        at at.forsyte.apalache.tla.Tool.main(Tool.scala)
```

I guess this is because we run typechecking which eats it here, before we do the `PreproPass`?

https://github.com/informalsystems/apalache/blob/aa87a154ea7e7ec59efb7df00fa8e628eddbf062/tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/config/CheckerModule.scala#L91-L102

**Comment by shonfeder** [comment]
I can just bypass this check in the typechecker, but then we abandon compat with TLC in our typing, and I agree with @konnov that this edge case doesn't justify it.

If we can't get this working with our current pipeline, probably best to revert.

**Comment by Kukovec** [comment]
> I can just bypass this check in the typechecker, but then we abandon compat with TLC in our typing, and I agree with @konnov that this edge case doesn't justify it.

I don't think you need to "bypass" anything, we just have to add a case for the typechecker to understand that `(LET F == ... IN F)` has an operator type, and continue from there. For the typechecker, `(LET F == ... IN F)(x)` and `F(x)` + in-scope definition should be equivalent, it's just currently not accepting the former on a syntax basis.

**Comment by shonfeder** [comment]
This would make the type checker accept TLA that is not valid with TLC. This is what I mean by “bypass”.

As per our discussion yesterday, it doesn’t seem think there’s any reason to deviate in this way, because we can get what we need from this change with just five lines in the scoped builder, and this is all just a workaround anyhow. :)

**Pull request merged**